File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/broadcasters/specs/GivenAPlayingFile/WhenBroadcastingTheFileProgress.java
package com.lasthopesoftware.bluewater.client.playback.service.broadcasters.specs.GivenAPlayingFile;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.v4.content.LocalBroadcastManager;

import com.lasthopesoftware.bluewater.client.playback.file.PlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.PlayedFile;
import com.lasthopesoftware.bluewater.client.playback.file.PlayingFile;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.TrackPositionBroadcaster;
import com.lasthopesoftware.bluewater.shared.promises.extensions.ProgressingPromise;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.lazyj.Lazy;

import org.joda.time.Duration;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import java.util.concurrent.CountDownLatch;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

@RunWith(RobolectricTestRunner.class)
public class WhenBroadcastingTheFileProgress {

	private static long progress;

	private static long duration;

	private static final Lazy<Void> setupTest = new Lazy<>(() -> {
		final LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.getInstance(RuntimeEnvironment.application);

		final CountDownLatch countDownLatch = new CountDownLatch(1);
		localBroadcastManager.registerReceiver(new BroadcastReceiver() {
			@Override
			public void onReceive(Context context, Intent intent) {
				duration = intent.getLongExtra(TrackPositionBroadcaster.TrackPositionChangedParameters.fileDuration,-1);
				progress = intent.getLongExtra(TrackPositionBroadcaster.TrackPositionChangedParameters.filePosition, -1);
				countDownLatch.countDown();
			}
		}, new IntentFilter(TrackPositionBroadcaster.trackPositionUpdate));

		final TrackPositionBroadcaster trackPositionBroadcaster = new TrackPositionBroadcaster(
			localBroadcastManager,
			new PlayingFile() {
				@Override
				public Promise<PlayableFile> promisePause() {
					return Promise.empty();
				}

				@Override
				public ProgressingPromise<Duration, PlayedFile> promisePlayedFile() {
					return new ProgressingPromise<Duration, PlayedFile>() {
						@Override
						public Duration getProgress() {
							return Duration.ZERO;
						}
					};
				}

				@Override
				public Duration getDuration() {
					return Duration.standardMinutes(3);
				}
			});

		trackPositionBroadcaster.accept(
			Duration
				.standardSeconds(2)
				.plus(Duration.standardSeconds(30)));

		countDownLatch.await();

		return null;
	});

	@Before
	public void before() {
		setupTest.getObject();
	}

	@Test
	public void thenTheProgressIsCorrect() {
		assertThat(progress).isEqualTo(Duration
			.standardSeconds(2)
			.plus(Duration.standardSeconds(30)).getMillis());
	}

	@Test
	public void thenTheDurationIsCorrect() {
		assertThat(duration).isEqualTo(Duration.standardMinutes(3).getMillis());
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/service/broadcasters/GivenAPlayingFile/WhenBroadcastingTheFileProgress.kt
package com.lasthopesoftware.bluewater.client.playback.service.broadcasters.GivenAPlayingFile

import com.lasthopesoftware.bluewater.client.playback.file.PlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.PlayedFile
import com.lasthopesoftware.bluewater.client.playback.file.PlayingFile
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.TrackPositionBroadcaster
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.messages.TrackPositionUpdate
import com.lasthopesoftware.promises.extensions.ProgressedPromise
import com.lasthopesoftware.promises.extensions.toPromise
import com.lasthopesoftware.resources.RecordingApplicationMessageBus
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.mockk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.joda.time.Duration
import org.junit.jupiter.api.Test

class WhenBroadcastingTheFileProgress {

	private val receivedMessage by lazy {
		val appMessageBus = RecordingApplicationMessageBus()
		val trackPositionBroadcaster = TrackPositionBroadcaster(appMessageBus, mockk())
		trackPositionBroadcaster.observeUpdates(object : PlayingFile {
			override fun promisePause(): Promise<PlayableFile> {
				return Promise.empty()
			}

			override fun promisePlayedFile(): ProgressedPromise<Duration, PlayedFile> {
				return object : ProgressedPromise<Duration, PlayedFile>() {
					override val progress: Promise<Duration>
						get() = Duration.ZERO.toPromise()
				}
			}

			override val duration: Promise<Duration>
				get() = Duration.standardMinutes(3).toPromise()

			override val progress: Promise<Duration>
				get() = promisePlayedFile().progress
		}).accept(
			Duration
				.standardSeconds(2)
				.plus(Duration.standardSeconds(30)))

		appMessageBus.recordedMessages.first() as? TrackPositionUpdate
	}

	private val duration by lazy {
		receivedMessage?.fileDuration
	}

	private val progress by lazy {
		receivedMessage?.filePosition
	}

	@Test
	fun `then the progress is correct`() {
		assertThat(progress).isEqualTo(Duration
			.standardSeconds(2)
			.plus(Duration.standardSeconds(30)))
	}

	@Test
	fun `then the duration is correct`() {
		assertThat(duration).isEqualTo(Duration.standardMinutes(3))
	}
}
